src/bootloader.rs

```diff
@@ -0,0 +1,500 @@
+        self.lines
+            .iter()
+            .filter_map(|line| strip_ble_key(line, key))
+    }
+
+    pub fn get_value(&self, key: &str) -> Option<&str> {
+        self.lines.iter().find_map(|line| strip_ble_key(line, key))
+    }
+
+    /// Adds a kernel command-line argument, possibly replacing a previous value.
+    ///
+    /// arg can be something like "composefs=xyz" but it can also be something like "rw".  In
+    /// either case, if the argument already existed, it will be replaced.
+    pub fn add_cmdline(&mut self, arg: &str) {
+        let key = match arg.find('=') {
+            Some(pos) => &arg[..=pos], // include the '='
+            None => arg,
+        };
+
+        // There are three possible paths in this function:
+        //   1. options line with key= already in it (replace it)
+        //   2. options line with no key= in it (append key=value)
+        //   3. no options line (append the entire thing)
+        for line in &mut self.lines {
+            if let Some(cmdline) = strip_ble_key(line, "options") {
+                let segment = split_cmdline(cmdline).find(|s| s.starts_with(key));
+
+                if let Some(old) = segment {
+                    // 1. Replace existing key
+                    let range = substr_range(line, old).unwrap();
+                    line.replace_range(range, arg);
+                } else {
+                    // 2. Append new argument
+                    line.push(' ');
+                    line.push_str(arg);
+                }
+
+                return;
+            }
+        }
+
+        // 3. Append new "options" line with our argument
+        self.lines.push(format!("options {arg}"));
+    }
+
+    /// Adjusts the kernel command-line arguments by adding a composefs= parameter (if appropriate)
+    /// and adding additional arguments, as requested.
+    pub fn adjust_cmdline(&mut self, composefs: Option<&str>, extra: &[&str]) {
+        if let Some(id) = composefs {
+            self.add_cmdline(&format!("composefs={id}"));
+        }
+
+        for item in extra {
+            self.add_cmdline(item);
+        }
+    }
+}
+
+// TODO: put this in a better spot
+// Reminder: it only makes sense to do boot stuff on a legit repository because we need to read the
+// boot loader entry content.  But: wouldn't it be nice to use the trait convenience wrappers
+// anyway?
+fn read_file<ObjectID: FsVerityHashValue>(
+    file: &RegularFile<ObjectID>,
+    repo: &Repository<ObjectID>,
+) -> Result<Box<[u8]>> {
+    match file {
+        RegularFile::Inline(data) => Ok(data.clone()),
+        RegularFile::External(id, size) => {
+            // TODO: mmap niceness?
+            // TODO: verify size: it could be faked
+            let mut data = Vec::with_capacity(*size as usize);
+            // TODO: read loop...
+            read(repo.open_object(id)?, spare_capacity(&mut data)).unwrap();
```

Comment:
Why the unwrap?

Reply:
"Move fast and do broken things".

I'll definitely fix this.  Thanks for the catch :)

Reply:
Cleaned this up by just using File::from().read_to_end() on the fd and comparing the length after.  It's a lot neater.